### PR TITLE
[RHCLOUD-18056] App reconiliations times available via prometheus

### PR DESCRIPTION
**Pull request by adamrdrew**
This MR adds instrumentation for app reconciliations times available via prometheus. The bulk of the work is in `controllers/cloud.redhat.com/clowdapp_controller.go` where 2 new reconciliation metrics methods and some state have been added to `ClowdAppReconciler`. A new metric has been registered with prometheus in `controllers/cloud.redhat.com/metrics.go` and new unit tests have been added via `controllers/cloud.redhat.com/clowdapp_controller_test.go`

This change exposes a new `clowder_app_reconciliation_time` histogram vec. It can be queried by app name and cloud env. Here's an example of the metrics output while running a kuttl test:

```
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="0.005"} 0
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="0.01"} 0
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="0.025"} 0
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="0.05"} 0
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="0.1"} 0
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="0.25"} 1
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="0.5"} 2
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="1"} 2
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="2.5"} 2
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="5"} 2
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="10"} 2
clowder_app_reconciliation_time_bucket{app="puptoo",env="test-basic-app",le="+Inf"} 2
clowder_app_reconciliation_time_sum{app="puptoo",env="test-basic-app"} 0.543926348
clowder_app_reconciliation_time_count{app="puptoo",env="test-basic-app"} 2
```

### CODE REVIEW

File: controllers/cloud.redhat.com/clowderconfig/config.go
Comment (psav) [inline controllers/cloud.redhat.com/clowderconfig/config.go:53]: This should not be changed - this is where clowder looks to get its config from in production

File: bundle/tests/scorecard/config.yaml
Comment (psav) [inline bundle/tests/scorecard/config.yaml:13]: These first three files seem to have been changed due to a newer operator-sdk version

File: config/manager/kustomization.yaml
Comment (psav) [inline config/manager/kustomization.yaml:21]: We ignore the changes to this file - it should be removed from the PR

File: bundle/tests/scorecard/config.yaml
Comment (psav) [reply bundle/tests/scorecard/config.yaml:13]: We should go through and ensure we are all running the same version - it's been a while since we did an upgrade - there are generally changes we need to make to the code base after an upgrade - we should get a card in place to do this

File: controllers/cloud.redhat.com/clowdapp_controller.go
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_controller.go:83]: My only concern here is that this struct becomes large with a bunch of new fields, is there a category we could put it under, almost everything else is really core/crucial,k any ideas?

File: controllers/cloud.redhat.com/clowdapp_controller.go
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_controller.go:105]: Generally the positive here, especially as the function is so small, generates a better readability

```
if clowderconfig.LoadedConfig.Features.ReconciliationMetrics {
	r.ReconcileStartTime = time.Now()
}
```

File: controllers/cloud.redhat.com/clowdapp_controller.go
Comment (psav) [inline controllers/cloud.redhat.com/clowdapp_controller.go:111]: Same as above

File: controllers/cloud.redhat.com/version.txt
Comment (psav) [inline controllers/cloud.redhat.com/version.txt:1]: This should be omitted

**Review by psav — CHANGES_REQUESTED** [review]

File: controllers/cloud.redhat.com/clowdapp_controller.go
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go:83]: I could factor reconciliation metrics out into its own struct and use that in `ClowdAppReconciler.` It could track its own state. That would be the cleaner design I think.

Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowdapp_controller.go]: Factored it out, see new commits after checks pass

File: controllers/cloud.redhat.com/clowderconfig/config.go
Comment (adamrdrew) [reply controllers/cloud.redhat.com/clowderconfig/config.go]: oops

File: bundle/tests/scorecard/config.yaml
Comment (psav) [reply bundle/tests/scorecard/config.yaml:13]: operator-sdk is pretty easy to get the old version - but we _should_ look at updating in clowder - can you get a card together for an operator-sdk update in clowder?

**Comment by adamrdrew** [comment]
@psav I opened RHCLOUD-18773 to track operator-sdk upgrade. What should I do here? I don't know how operator-sdk got upgraded and I don't know how to change it.

**Comment by psav** [comment]
This one is almost ready to go - I just noticed one thing

``config/deployment-template/clowder_config.yaml`` - needs updating with the new flag
``template.yml`` - needs updating to set it to false by default

**Review by psav — APPROVED** [review]

**Comment by adamrdrew** [comment]
We discussed this before the holiday - I think we're good to go on this? We went over a lot of stuff but I didn't take any notes on this one - my bad if I missed any TODOs on my part. IIRC we didn't want to merge it before the holiday, and then since we've had problems in ephemeral so we're probably not looking to merge. Please let me know if there's anything I missed on this, otherwise I think we're just waiting on merge.

**Comment by psav** [comment]
I'd love to merge this - can we get a rebase done?

**Comment by adamrdrew** [comment]
@psav rebase done - after checks pass feel free to merge!

**Comment by psav** [comment]
Gah - needs another rebase :(

**Comment by adamrdrew** [comment]
Tried again - if this isn't right then something is wrong with my rebase procedure and I'll request some help with that.

**Comment by adamrdrew** [comment]
Manual rebase via patch on master and force push. Hope this is alright! Thanks for the help.

**Review by BlakeHolifield — APPROVED** [review]

**Pull request merged**
